server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,228 @@
+      livyConf: LivyConf): SparkYarnApp =
+    new SparkYarnApp(appTag, None, process, listener, livyConf)
+
+  private def getYarnTagToAppIdTimeout(livyConf: LivyConf): FiniteDuration =
+    livyConf.getTimeAsMs(LivyConf.YARN_APP_LOOKUP_TIMEOUT) milliseconds
+
+  private def getYarnPollInterval(livyConf: LivyConf): FiniteDuration =
+    livyConf.getTimeAsMs(LivyConf.YARN_POLL_INTERVAL) milliseconds
+}
+
+/**
+ * Provide a class to control a Spark application using YARN API.
+ *
+ * @param appId A function that returns the YARN application id for this application.
+ * @param process The spark-submit process launched the YARN application. This is optional.
+ *                If it's provided, SparkYarnApp.log() will include its log.
+ */
+class SparkYarnApp private[utils] (
+    appTag: String,
+    appIdOption: Option[ApplicationId],
+    process: Option[LineBufferedProcess],
+    listener: Option[SparkAppListener],
+    livyConf: LivyConf,
+    yarnClient: => YarnClient = SparkYarnApp.yarnClient) // For unit test.
+  extends SparkApp
+  with Logging {
+  import SparkYarnApp._
+
+  private val appIdPromise: Promise[ApplicationId] = Promise()
+  private var state: SparkApp.State = SparkApp.State.STARTING
+  private var yarnDiagnostics: IndexedSeq[String] = IndexedSeq.empty[String]
+
+  override def log(): IndexedSeq[String] =
+    process.map(_.inputLines).getOrElse(ArrayBuffer.empty[String]) ++ yarnDiagnostics
+
+  override def kill(): Unit = synchronized {
+    if (isRunning) {
+      try {
+        val timeout = SparkYarnApp.getYarnTagToAppIdTimeout(livyConf)
+        yarnClient.killApplication(Await.result(appIdPromise.future, timeout))
+      } catch {
+        // We cannot kill the YARN app without the app id.
+        // There's a chance the YARN app hasn't been submitted during a livy-server failure.
+        // We don't want a stuck session that can't be deleted. Emit a warning and move on.
+        case _: TimeoutException | _: InterruptedException =>
+          warn("Deleting a session while its YARN application is not found.")
+          yarnAppMonitorThread.interrupt()
+      } finally {
+        process.foreach(_.destroy())
+      }
+    }
+  }
+
+  private def changeState(newState: SparkApp.State.Value): Unit = {
+    if (state != newState) {
+      listener.foreach(_.stateChanged(state, newState))
+      state = newState
+    }
+  }
+
+  /**
+   * Find the corresponding YARN application id from an application tag.
+   *
+   * @param appTag The application tag tagged on the target application.
+   *               If the tag is not unique, it returns the first application it found.
+   *               It will be converted to lower case to match YARN's behaviour.
+   * @return ApplicationId or the failure.
+   */
+  @tailrec
+  private def getAppIdFromTag(
+    appTag: String,
+    pollInterval: Duration = getYarnPollInterval(livyConf),
+    deadline: Deadline = getYarnTagToAppIdTimeout(livyConf).fromNow): ApplicationId = {
```

Comment:
I see this one can't be a local variable; but I'd prefer an explicit parameter instead of one with a default value.